src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterKeyValuePair.cs

```diff
@@ -3,22 +3,26 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Generic;
-using System.Diagnostics.CodeAnalysis;
 
 namespace System.Text.Json.Serialization.Converters
 {
-    internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonConverter<KeyValuePair<TKey, TValue>>
+    internal sealed class JsonKeyValuePairConverter<TKey, TValue> : JsonValueConverter<KeyValuePair<TKey, TValue>>
     {
         private const string KeyName = "Key";
         private const string ValueName = "Value";
 
-        // "encoder: null" is used since the literal values of "Key" and "Value" should not normally be escaped
-        // unless a custom encoder is used that escapes these ASCII characters (rare).
-        // Also by not specifying an encoder allows the values to be cached statically here.
+        // todo: move these to JsonSerializerOptions and use the proper encoding.
         private static readonly JsonEncodedText _keyName = JsonEncodedText.Encode(KeyName, encoder: null);
         private static readonly JsonEncodedText _valueName = JsonEncodedText.Encode(ValueName, encoder: null);
 
-        public override KeyValuePair<TKey, TValue> Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        // todo: it is possible to cache the underlying converters since this is an internal converter and
+        // an instance is created only once for each JsonSerializerOptions instance.
+
+        internal override bool OnTryRead(
```

Comment:
I'm going to revisit this once I spend more time on the public API. Likely it won't be OnXXX. 

The idea was a single pattern to override (OnXXX) for both value converters and object\collection converters and let the converter that can actually run out of data or need to flush.

However after thinking about this for a few days I think it is best to not expose that level of control publically (i.e. we'll do the read-ahead for public object\collection converters similar to how we do it today internally meaning per-property and per-element) and thus the converters won't need to worry about it (it is too hard to manage the state transitions and continuation manually).

Reply:
Could `JsonValueConverterKeyValuePair ` override the Read/Write methods **instead** of OnTryRead/OnTryWrite?

Reply:
It could but it doesn't since this was changed to use the new re-entry method that takes the additional ReadStack\WriteStack parameter and string property name which is only passed in the OnXXX methods.

Also this class will be re-written to just forward to the converters for perf. At that time, the re-entry methods should be public so tests can be written for those. Currently this converter is the only "test" for that so I didn't want to change it yet to forward to the converters.